File: leetcode/201-300/242-valid-anagram/bench/valid_anagram.c
/* Benchmark harness for LeetCode #242 — Valid Anagram.
 * Mirrors valid_anagram.kara algorithm-for-algorithm.
 *
 * No hash map here — the kata is a fixed 26-slot frequency array, so every
 * language runs the same structure. That makes this one of the cleaner
 * like-for-like comparisons in the corpus: no data-structure divergence to
 * caveat, unlike #290/#291 where C's map differed from the others.
 */

#include <stdio.h>
#include <string.h>

#define NP 8
#define SL 20000
#define ITERS 8000

static unsigned char esses[NP][SL];
static unsigned char tees[NP][SL];

static int is_anagram(const unsigned char *s, const unsigned char *t, long long len) {
    long long count[26];
    memset(count, 0, sizeof(count));
    for (long long i = 0; i < len; i++) {
        count[s[i] - 97] += 1;
        count[t[i] - 97] -= 1;
    }
    for (int j = 0; j < 26; j++) {
        if (count[j] != 0) {
            return 0;
        }
    }
    return 1;
}

int main(void) {
    for (long long j = 0; j < NP; j++) {
        for (long long k = 0; k < SL; k++) {
            esses[j][k] = (unsigned char)(97 + ((k * 7 + j) % 26));
        }
        long long w = 0;
        for (long long m = SL - 1; m >= 0; m--) {
            long long b = esses[j][m];
            if (j % 2 == 1 && m == 0) {
                b = 97 + ((b - 97 + 1) % 26);
            }
            tees[j][w++] = (unsigned char)b;
        }
    }

    long long sink = 0;
    for (long long it = 0; it < ITERS; it++) {
        long long idx = (it * 3) % NP;
        if (is_anagram(esses[idx], tees[idx], SL)) {
            sink += it + 1;
        } else {
            sink += 1;
        }
    }
    printf("%lld\n", sink);
    return 0;
}

File: leetcode/201-300/242-valid-anagram/bench/valid_anagram.rs
// Benchmark harness for LeetCode #242 — Valid Anagram.
// Mirrors valid_anagram.kara algorithm-for-algorithm.

fn is_anagram(s: &[u8], t: &[u8]) -> bool {
    if s.len() != t.len() {
        return false;
    }
    let mut count = [0i64; 26];
    for i in 0..s.len() {
        count[(s[i] as usize) - 97] += 1;
        count[(t[i] as usize) - 97] -= 1;
    }
    for j in 0..26 {
        if count[j] != 0 {
            return false;
        }
    }
    true
}

fn main() {
    let np: i64 = 8;
    let sl: i64 = 20000;
    let iters: i64 = 8000;

    let mut esses: Vec<Vec<u8>> = Vec::new();
    let mut tees: Vec<Vec<u8>> = Vec::new();
    for j in 0..np {
        let mut sj: Vec<u8> = Vec::new();
        for k in 0..sl {
            sj.push((97 + ((k * 7 + j) % 26)) as u8);
        }
        let mut tj: Vec<u8> = Vec::new();
        let mut m = sl - 1;
        while m >= 0 {
            let mut b = sj[m as usize] as i64;
            if j % 2 == 1 && m == 0 {
                b = 97 + ((b - 97 + 1) % 26);
            }
            tj.push(b as u8);
            m -= 1;
        }
        esses.push(sj);
        tees.push(tj);
    }

    let mut sink: i64 = 0;
    for it in 0..iters {
        let idx = ((it * 3) % np) as usize;
        if is_anagram(&esses[idx], &tees[idx]) {
            sink += it + 1;
        } else {
            sink += 1;
        }
    }
    println!("{}", sink);
}
